### PR TITLE
Do not crash on lexically bad floating-point tokens

### LINKED ISSUE #6079: FormatException when typing "0" (zero) in incomplete statement
capture

Dump at:  [internal file share]\6079\FormatException.dmp
```
System.FormatException
>	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.ErrorReporting.FatalError.Report(System.Exception exception, System.Action<System.Exception> handler) + 0x38 bytes	
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.DocumentState.IncrementallyParseTreeAsync() + 0x5fb bytes	
 	mscorlib.dll!System.Number.StringToNumber(string str, System.Globalization.NumberStyles options, ref System.Number.NumberBuffer number, System.Globalization.NumberFormatInfo info, bool parseDecimal) + 0xed bytes	
 	mscorlib.dll!System.Number.ParseInt32(string s, System.Globalization.NumberStyles style, System.Globalization.NumberFormatInfo info) + 0x79 bytes	
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.RealParser.DecimalFloatingPointString.FromSource(string source) + 0x1a6 bytes	
 	Microsoft.CodeAnalysis.dll!Microsoft.CodeAnalysis.RealParser.TryParseDouble(string s, out double d) + 0x21 bytes	
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer.GetValueDouble(string text) + 0x1e bytes	
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer.ScanNumericLiteral(ref Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer.TokenInfo info) + 0x522 bytes	
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer.ScanSyntaxToken(ref Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer.TokenInfo info) + 0x5fe bytes	
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer.LexSyntaxToken() + 0x69 bytes	
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Lexer.Lex(Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.LexerMode mode) + 0x2f bytes	
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Blender.Reader.LexNewToken(Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.LexerMode mode) + 0x5b bytes	
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Blender.Reader.ReadNewToken(Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.LexerMode mode) + 0x15 bytes	
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Blender.Reader.ReadNodeOrToken(Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.LexerMode mode, bool asToken) + 0x6d bytes	
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.Blender.ReadNodeOrToken(Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.LexerMode mode, bool asToken) + 0x8a bytes	
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SyntaxParser.AddNewToken() + 0x73 bytes	
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SyntaxParser.FetchCurrentToken() + 0x16 bytes	
 	Microsoft.CodeAnalysis.CSharp.dll!Microsoft.CodeAnalysis.CSharp.Syntax.InternalSyntax.SyntaxParser.CurrentToken.get() + 0x13 bytes
```

### PR DESCRIPTION
**Pull request by gafter**
Fixes #6079 

@dotnet/roslyn-compiler Please review.

### CODE REVIEW

**Comment by AlekseyTs** [comment]
LGTM

**Comment by gafter** [comment]
@dotnet/roslyn-compiler One more review, please? This is a small change.

File: src/Compilers/Core/Portable/RealParser.cs [hunk 1]
```diff
@@ -47,12 +56,20 @@ public static bool TryParseDouble(string s, out double d)
...
+            try
+            {
+                var str = DecimalFloatingPointString.FromSource(s);
+                var dbl = FloatFloatingPointType.Instance;
+                ulong result;
+                var status = RealParser.ConvertDecimalToFloatingPointBits(str, dbl, out result);
+                f = Int32BitsToFloat((uint)result);
+                return status != Status.Overflow;
+            }
+            catch (System.FormatException)
```
Comment (davkean) [inline src/Compilers/Core/Portable/RealParser.cs]: Which of the methods actually leaks the FormatException? You've got 4 method calls in that try block, and they could all FormatExceptoon for other reasons - like an incorrectly formatted string.

Comment (gafter) [inline src/Compilers/Core/Portable/RealParser.cs]: It occurs in `DecimalFloatingPointString.FromSource` on a string such as `"1.0e"` where the exponent is empty. See also related #6109.

**Comment by TyOverby** [comment]
LGTM

**Pull request merged**